src/NHibernate/Driver/OracleDataClientDriverBase.cs

```diff
@@ -0,0 +1,171 @@
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Data.Common;
+using System.Reflection;
+using NHibernate.AdoNet;
+using NHibernate.Engine.Query;
+using NHibernate.SqlTypes;
+using NHibernate.Util;
+
+namespace NHibernate.Driver
+{
+	/// <summary>
+	/// A NHibernate driver base for using ODP.Net.
+	/// </summary>
+	/// <remarks>
+	/// Original code was contributed by <a href="http://sourceforge.net/users/jemcalgary/">James Mills</a>
+	/// on the NHibernate forums in this 
+	/// <a href="http://sourceforge.net/forum/message.php?msg_id=2952662">post</a>.
+	/// </remarks>
+	public abstract class OracleDataClientDriverBase : ReflectionBasedDriver, IEmbeddedBatcherFactoryProvider
+	{
+		private const string _commandClassName = "OracleCommand";
+
+		private static readonly SqlType _guidSqlType = new SqlType(DbType.Binary, 16);
+
+		private readonly Action<object, bool> _commandBindByNameSetter;
+		private readonly Action<object, object> _parameterOracleDbTypeSetter;
+		private readonly object _oracleDbTypeRefCursor;
+		private readonly object _oracleDbTypeXmlType;
+		private readonly object _oracleDbTypeBlob;
+		private readonly object _oracleDbTypeNVarchar2;
+		private readonly object _oracleDbTypeNChar;
+
+		/// <summary>
+		/// Default constructor.
+		/// </summary>
+		/// <param name="assemblyName">The assembly name of the managed or unmanage driver. Namespaces will be derived from it.</param>
+		/// <exception cref="HibernateException">
+		/// Thrown when the requested assembly can not be loaded.
+		/// </exception>
+		protected OracleDataClientDriverBase(string assemblyName)
+			: this(assemblyName, assemblyName + ".Client")
+		{
+		}
+
+		private OracleDataClientDriverBase(string driverAssemblyName, string clientNamespace)
+			: base(clientNamespace, driverAssemblyName, clientNamespace + ".OracleConnection", clientNamespace + "." + _commandClassName)
+		{
+			var oracleCommandType = ReflectHelper.TypeFromAssembly(clientNamespace + "." + _commandClassName, driverAssemblyName, true);
+			_commandBindByNameSetter = DelegateHelper.BuildPropertySetter<bool>(oracleCommandType, "BindByName");
+
+			var parameterType = ReflectHelper.TypeFromAssembly(clientNamespace + ".OracleParameter", driverAssemblyName, true);
+			_parameterOracleDbTypeSetter = DelegateHelper.BuildPropertySetter<object>(parameterType, "OracleDbType");
```

Comment:
We can not use the actual set type here for the delegate. This has required adapting `DelegateHelper`.